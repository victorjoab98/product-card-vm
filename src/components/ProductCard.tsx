import  React,{ createContext } from 'react'
import styles from '../styles/styles.module.css'
import { useProduct } from '../hooks'
import { InitialValuesProductCard, onChangeProductArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product,
  // children?: ReactElement | ReactElement[],
  children: ( args: ProductCardHandlers ) => JSX.Element,
  className?: string,
  style?: React.CSSProperties,
  onChange?: ( args: onChangeProductArgs ) => void,
  value?: number,
  initialValues?: InitialValuesProductCard
}

export const ProductCard = ( { product, children, className, style, onChange, value, initialValues } : Props ) => {

  const { counter, increaseBy, reset, maxCount, isMaxCountReached } = useProduct({ product, onChange, value, initialValues })

  return (
    <Provider value={{
        product,
        counter,
        increaseBy,
        maxCount
    }}> 
        <div className={ `${ styles.productCard } ${ className }` } style={style}>
            { children({
                count: counter,
                isMaxCountReached,
                maxCount,
                product,
                increaseBy,
                reset
            }) }
        </div>
    </Provider>
  )
}
