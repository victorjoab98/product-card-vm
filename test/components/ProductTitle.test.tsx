import React from 'react';
import renderer from 'react-test-renderer';
import ProductCard, { ProductTitle } from '../../src/components';
import { product1 } from '../data/products';

describe('ProducTitle', () => {
    test('debe de mostrar el componente con el titulo personalizado', () => {
        const wrapper = renderer.create(
            <ProductTitle title='Hola mundo'/>
        )
        expect( wrapper.toJSON() ).toMatchSnapshot();
    })

    test('debe mostrar el titulo que viene del componenten padre', () => {

        const wrapper = renderer.create(
            <ProductCard  product={product1}>
                {
                    () => (
                        <ProductTitle/>
                    )
                }
            </ProductCard>
        )

        expect( wrapper.toJSON() ).toMatchSnapshot();

    })
})