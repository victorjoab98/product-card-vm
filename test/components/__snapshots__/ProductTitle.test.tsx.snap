// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProducTitle debe de mostrar el componente con el titulo personalizado 1`] = `
<span
  className="productDescription undefined"
>
  Hola mundo
</span>
`;

exports[`ProducTitle debe mostrar el titulo que viene del componenten padre 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffe Mug - Con Imagen
  </span>
</div>
`;
